# IND320 Project work 1

**Autor:** Inge Gilje Gunnarshaug

## 1. Introduction

This is the notebook for the first task in IND320. The corresponding streamlit app and github repository can be found here:

### Project links

- **GitHub repository:** https://github.com/IngeGiljeGunnarshaug/ind320-inge-gilje-gunnarshaug
- **Streamlit app:** https://ind320-inge-gilje-gunnarshaug.streamlit.app/

## 2. Environment and Imports

This notebook uses Pandas for data processing and Matplotlib for plotting.
The CSV file is stored at `data/reservoirs.csv`, relative to this notebook.

The required packages and tested versions are documented in the repository.

In [5]:
from pathlib import Path

import pandas as pd
import matplotlib.pyplot as plt

# Path relative to the notebook's working directory.
DATA_PATH = Path("data") / "reservoirs.csv"

# Show the working directory to help diagnose file-path problems.
print("Working directory:", Path.cwd())

if not DATA_PATH.is_file():
    raise FileNotFoundError(f"CSV file not found: {DATA_PATH.resolve()}")

Working directory: /Users/ingegunnarshaug/Skole/H2026/ind320/ind320-inge-gilje-gunnarshaug


## 3. Load and Inspect the Data

### 3.1 Load data

In [6]:
# Preserve the imported data so preparation can be compared with the source.
raw_df = pd.read_csv(DATA_PATH)

print(f"Loaded {raw_df.shape[0]:,} rows and {raw_df.shape[1]} columns.")

Loaded 14,877 rows and 11 columns.


### 3.2 Preview and sizes

In [7]:
# These previews show the original file order, before chronological sorting.
print("First five rows:")
display(raw_df.head())

print("Last five rows:")
display(raw_df.tail())

print("Dataset dimensions (rows, columns):", raw_df.shape)

First five rows:


,dato_Id,omrType,omrnr,iso_aar,iso_uke,fyllingsgrad,kapasitet_TWh,fylling_TWh,neste_Publiseringsdato,fyllingsgrad_forrige_uke,endring_fyllingsgrad
0,1995-09-03,EL,4,1995,35,0.944721,21.079208,19.913979,0001-01-01T00:00:00,0.936888,0.007833
1,2020-11-15,EL,1,2020,46,0.974361,6.003264,5.849344,2020-11-25T13:00:00,0.997406,-0.023046
2,2011-08-28,EL,4,2011,34,0.786499,21.079208,16.578772,0001-01-01T00:00:00,0.787193,-0.000694
3,1998-07-05,EL,3,1998,27,0.793969,8.920932,7.082947,0001-01-01T00:00:00,0.733812,0.060157
4,2011-03-27,EL,4,2011,12,0.300820,21.079208,6.341054,0001-01-01T00:00:00,0.311122,-0.010301


Last five rows:


,dato_Id,omrType,omrnr,iso_aar,iso_uke,fyllingsgrad,kapasitet_TWh,fylling_TWh,neste_Publiseringsdato,fyllingsgrad_forrige_uke,endring_fyllingsgrad
14872,2026-07-19,VASS,3,2026,29,0.796849,28.155403,22.435612,2026-07-29T13:00:00,0.794957,0.001892
14873,2026-09-06,VASS,3,2026,36,0.826252,28.155403,23.263466,2026-09-16T13:00:00,0.827641,-0.001389
14874,2026-08-30,VASS,3,2026,35,0.827718,28.155403,23.304743,2026-09-09T13:00:00,0.829898,-0.002180
14875,2026-09-06,VASS,1,2026,36,0.570397,36.044464,20.559645,2026-09-16T13:00:00,0.567087,0.003310
14876,2026-09-06,VASS,2,2026,36,0.516978,23.237715,12.013381,2026-09-16T13:00:00,0.514423,0.002555


Dataset dimensions (rows, columns): (14877, 11)


11 columns with norwegian headers, with different datatype; some numerical, some categorical and some dates. Seems like 0001-01-01T00:00:00 is a placeholder for missing neste_Publiseringsdato. It is also worth notecing that the rows are not sorted by date.

### 3.3 datatypes and summaries

In [8]:
# Inspect inferred types and the number of nonmissing values.
raw_df.info()

# Numeric summaries include identifiers, so interpret them accordingly.
print("\nNumeric summary statistics:")
display(raw_df.describe().T)

# Inspect text columns separately.
print("Text-column summary:")
display(raw_df.describe(include=["object", "string"]).T)

<class 'pandas.DataFrame'>
RangeIndex: 14877 entries, 0 to 14876
Data columns (total 11 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   dato_Id                   14877 non-null  str    
 1   omrType                   14877 non-null  str    
 2   omrnr                     14877 non-null  int64  
 3   iso_aar                   14877 non-null  int64  
 4   iso_uke                   14877 non-null  int64  
 5   fyllingsgrad              14877 non-null  float64
 6   kapasitet_TWh             14877 non-null  float64
 7   fylling_TWh               14877 non-null  float64
 8   neste_Publiseringsdato    14877 non-null  str    
 9   fyllingsgrad_forrige_uke  14877 non-null  float64
 10  endring_fyllingsgrad      14877 non-null  float64
dtypes: float64(5), int64(3), str(3)
memory usage: 1.7 MB

Numeric summary statistics:


,count,mean,std,min,25%,50%,75%,max
omrnr,14877.0,2.333333,1.490762,0.000000,1.000000,2.000000,3.000000,5.000000
iso_aar,14877.0,2010.346038,9.146696,1995.000000,2002.000000,2010.000000,2018.000000,2026.000000
iso_uke,14877.0,26.405929,15.017910,1.000000,13.000000,26.000000,39.000000,53.000000
fyllingsgrad,14877.0,0.618852,0.214850,0.055160,0.456234,0.648049,0.802135,1.020072
kapasitet_TWh,14877.0,29.145860,22.739070,6.003264,17.390587,23.237715,34.043590,87.437580
fylling_TWh,14877.0,18.139713,15.966211,0.331142,7.321973,14.501389,22.203530,84.544820
fyllingsgrad_forrige_uke,14877.0,0.618890,0.214843,0.055160,0.456356,0.648114,0.802133,1.020037
endring_fyllingsgrad,14877.0,-0.000038,0.031385,-0.242484,-0.022395,-0.007072,0.016277,0.336623


Text-column summary:


,count,unique,top,freq
dato_Id,14877,1653,1995-09-03,9
omrType,14877,3,EL,8265
neste_Publiseringsdato,14877,396,0001-01-01T00:00:00,11322


The numerical columns seems to have different scales, varying from (-0.25, 0.33) to (0.3, 85). \
neste_Publiseringsdate has 11322 observations with value 0001-01-01T00:00:00.

### 3.4 Missing values, duplicates and area identifiers

In [16]:
# Blank CSV fields may become missing values; placeholder strings do not.
missing_summary = pd.DataFrame({
    "missing_count": raw_df.isna().sum(),
    "missing_percent": raw_df.isna().mean() * 100,
})

print("Missing values before preparation:")
display(missing_summary)

# Repeated dates across different areas are expected.
# A duplicate observation has the same date AND both area identifiers.
raw_key = ["dato_Id", "omrType", "omrnr"]

print("Fully duplicated rows:", raw_df.duplicated().sum())
print("Duplicated observation keys:", raw_df.duplicated(subset=raw_key).sum())

# Area numbers must be interpreted together with their area type.
area_summary = (
    raw_df.groupby(["omrType", "omrnr"])
    .size()
    .reset_index(name="observation_count")
    .sort_values(["omrType", "omrnr"])
)

print("\nAvailable areas:")
display(area_summary)

Missing values before preparation:


,missing_count,missing_percent
dato_Id,0,0.0
omrType,0,0.0
omrnr,0,0.0
iso_aar,0,0.0
iso_uke,0,0.0
fyllingsgrad,0,0.0
kapasitet_TWh,0,0.0
fylling_TWh,0,0.0
neste_Publiseringsdato,0,0.0
fyllingsgrad_forrige_uke,0,0.0


Fully duplicated rows: 0
Duplicated observation keys: 0

Available areas:


,omrType,omrnr,observation_count
0,EL,1,1653
1,EL,2,1653
2,EL,3,1653
3,EL,4,1653
4,EL,5,1653
5,NO,0,1653
6,VASS,1,1653
7,VASS,2,1653
8,VASS,3,1653


No missing values before any preperations. Also no duplicates found. \
All the area identifiers have exactly the same number of observations.

## 4. Column Definitions and Data Preparation

### 4.1 Rename columns


In [17]:
# Use descriptive names and retain units in the energy-column names.
column_mapping = {
    "dato_Id": "date",
    "omrType": "area_type",
    "omrnr": "area_number",
    "iso_aar": "iso_year",
    "iso_uke": "iso_week",
    "fyllingsgrad": "fill_fraction",
    "kapasitet_TWh": "capacity_twh",
    "fylling_TWh": "stored_energy_twh",
    "neste_Publiseringsdato": "next_publication_date",
    "fyllingsgrad_forrige_uke": "previous_week_fill_fraction",
    "endring_fyllingsgrad": "fill_fraction_change",
}

# Work on a separate DataFrame, keeping raw_df unchanged.
df = raw_df.rename(columns=column_mapping).copy()

display(pd.DataFrame({
    "original_name": list(column_mapping.keys()),
    "english_name": list(column_mapping.values()),
}))

,original_name,english_name
0,dato_Id,date
1,omrType,area_type
2,omrnr,area_number
3,iso_aar,iso_year
4,iso_uke,iso_week
5,fyllingsgrad,fill_fraction
6,kapasitet_TWh,capacity_twh
7,fylling_TWh,stored_energy_twh
8,neste_Publiseringsdato,next_publication_date
9,fyllingsgrad_forrige_uke,previous_week_fill_fraction


### 4.2 Parse dates, and handle placeholder

In [18]:
# Observation dates are essential: stop if a value cannot be parsed.
df["date"] = pd.to_datetime(
    df["date"],
    format="%Y-%m-%d",
    errors="raise",
)

# Treat the year-0001 publication timestamp as an unavailable-date placeholder.
# This is an explicit preparation assumption, not a missing observation date.
publication_text = df["next_publication_date"].astype("string")
placeholder_mask = publication_text.eq("0001-01-01T00:00:00").fillna(False)

df["next_publication_date"] = pd.to_datetime(
    publication_text.mask(placeholder_mask),
    format="%Y-%m-%dT%H:%M:%S",
    errors="raise",
)

print(f"Publication-date placeholders replaced: {placeholder_mask.sum():,}")
print("Observation period:",
    df["date"].min().date(),
    "to",
    df["date"].max().date(),
)

Publication-date placeholders replaced: 11,322
Observation period: 1995-01-08 to 2026-09-06


### 4.3 Sort rows by date.

In [19]:
# Group each area's observations together and order them by date.
df = (
    df.sort_values(["area_type", "area_number", "date"])
    .reset_index(drop=True)
)

display(df.head())

,date,area_type,area_number,iso_year,iso_week,fill_fraction,capacity_twh,stored_energy_twh,next_publication_date,previous_week_fill_fraction,fill_fraction_change
0,1995-01-08,EL,1,1995,1,0.621467,6.003264,3.730831,NaT,0.734888,-0.113421
1,1995-01-15,EL,1,1995,2,0.583100,6.003264,3.500502,NaT,0.621467,-0.038367
2,1995-01-22,EL,1,1995,3,0.550604,6.003264,3.305419,NaT,0.583100,-0.032496
3,1995-01-29,EL,1,1995,4,0.508440,6.003264,3.052297,NaT,0.550604,-0.042164
4,1995-02-05,EL,1,1995,5,0.469223,6.003264,2.816870,NaT,0.508440,-0.039217


### 4.4 Distinguish measurements from metadata, and apply numerical types

In [20]:
# Only these columns represent reservoir measurements.
measurement_columns = [
    "fill_fraction",
    "capacity_twh",
    "stored_energy_twh",
    "previous_week_fill_fraction",
    "fill_fraction_change",
]

metadata_columns = [
    "date",
    "area_type",
    "area_number",
    "iso_year",
    "iso_week",
    "next_publication_date",
]

# Validate that measurement values are numeric; unexpected text raises an error.
df[measurement_columns] = df[measurement_columns].apply(
    pd.to_numeric,
    errors="raise",
)

# Use these labels and multipliers in later plots.
# The underlying DataFrame retains the original fractions.
plot_settings = {
    "fill_fraction": {
        "label": "Reservoir fill (%)",
        "multiplier": 100,
    },
    "capacity_twh": {
        "label": "Reservoir capacity (TWh)",
        "multiplier": 1,
    },
    "stored_energy_twh": {
        "label": "Stored energy (TWh)",
        "multiplier": 1,
    },
    "previous_week_fill_fraction": {
        "label": "Previous week's fill (%)",
        "multiplier": 100,
    },
    "fill_fraction_change": {
        "label": "Weekly fill change (percentage points)",
        "multiplier": 100,
    },
}

# Display a compact description of the measurement units.
display(pd.DataFrame([
    {
        "column": column,
        "plot_label": settings["label"],
        "display_multiplier": settings["multiplier"],
    }
    for column, settings in plot_settings.items()
]))

,column,plot_label,display_multiplier
0,fill_fraction,Reservoir fill (%),100
1,capacity_twh,Reservoir capacity (TWh),1
2,stored_energy_twh,Stored energy (TWh),1
3,previous_week_fill_fraction,Previous week's fill (%),100
4,fill_fraction_change,Weekly fill change (percentage points),100


### 4.5 Select an area for analysis

In [21]:
# Choose a type-number combination from the area summary in section 3.4.
SELECTED_AREA_TYPE = "NO"
SELECTED_AREA_NUMBER = 0

# Filter using both identifiers to avoid mixing different area definitions.
area_mask = (
    df["area_type"].eq(SELECTED_AREA_TYPE)
    & df["area_number"].eq(SELECTED_AREA_NUMBER)
)

area_df = (
    df.loc[area_mask]
    .sort_values("date")
    .reset_index(drop=True)
    .copy()
)

if area_df.empty:
    raise ValueError("The selected area has no observations. Check section 3.4.")

area_label = f"{SELECTED_AREA_TYPE} — area {SELECTED_AREA_NUMBER}"

print("Selected area:", area_label)
print(f"Observations: {len(area_df):,}")
display(area_df.head())

Selected area: NO — area 0
Observations: 1,653


,date,area_type,area_number,iso_year,iso_week,fill_fraction,capacity_twh,stored_energy_twh,next_publication_date,previous_week_fill_fraction,fill_fraction_change
0,1995-01-08,NO,0,1995,1,0.608274,87.43758,53.185986,NaT,0.752703,-0.144430
1,1995-01-15,NO,0,1995,2,0.582372,87.43758,50.921207,NaT,0.608274,-0.025902
2,1995-01-22,NO,0,1995,3,0.562412,87.43758,49.175980,NaT,0.582390,-0.019977
3,1995-01-29,NO,0,1995,4,0.534508,87.43758,46.736122,NaT,0.562412,-0.027904
4,1995-02-05,NO,0,1995,5,0.506486,87.43758,44.285900,NaT,0.534489,-0.028003


## 5. Individual Variable Plots


## 6. Comparing Measurements with Different Scales


## 7. Streamlit App


## 8. Work Log

*300–500 words covering both Jupyter Notebook and Streamlit experience.*

Describe:
- What you implemented.
- How you approached data preparation and visualisation.
- How you developed and deployed the Streamlit app.
- Problems encountered and how you resolved them.
- What you learned and what you would improve.

## 9. AI Usage

Describe which AI tools you used, what they helped with, and how you
checked and adapted their suggestions.

